backend/controllers/AuthController.js

```diff
@@ -1 +1,318 @@
+
+    res.status(200).json({
+      message: " ><> Login successful. <><",
+      token,
+    });
+  } catch (error) {
+    // Handles invalid login attempts with appropriate error messages.
+    console.error("Error during login:", error);
+    res.status(500).json({ error: "Internal server error." });
+  }
+}
+
+/* 4. Log out a user: Invalidates tokens to ensure users can securely log out. */
+export async function logout(req, res) {
+  // TODO EXTRA: Involve token invalidation on multiple devices.
+  try {
+    const { userId } = req.user;
+    console.log("Incoming logout request:", userId);
+
+    // Find the user by id
+    const user = await UserModel.getUserById(userId);
+    console.log("Searching for user with Id...");
+    if (!user) {
+      return res.status(404).json({ error: "User not found." });
+    }
+
+    // Invalid current token
+    await UserModel.incrementTokenVersion(userId); 
+    console.log("Token incremented.");
+
+    res.status(200).json({ message: " ><> Logout successful. <><" });
+  } catch (error) {
+    // Handles invalid login attempts with appropriate error messages.
+    console.error("Error during logout:", error);
+    res.status(500).json({ error: "Internal server error." });
+  }
+}
+
+/* 5. Password reset: Allows user to reset password securely with a email verification. */
+export async function requestPasswordReset(req, res)  {
+  // Temporary reset token for email validation
+  try {
+    // Find the user by email
+    const { email } = req.body;
+    if (!email){
+      console.log("Email is missing.");
+      return res.status(400).json({ error: "Email is required." });
+    }
+
+    const user = await UserModel.getUserByEmail(email);
+    console.log("Finding user by email...");
+
+    if (!user) {
+      return res.status(404).json({ error: "User not found." })
+    }
+
+    console.log("Generating resetPassToken...");
+    const resetPassToken = crypto.randomBytes(32).toString("hex");
+    const tokenExpires = new Date(Date.now() + 3600000);  // 1h
+    await user.save();
+    console.log("Reset Password Token generated");
+
+    // Send reset email for verification
+    const transporter = nodemailer.createTransport({
+      service: "gmail",               // TODO Example using Gmail, is adjustable.
+      auth: {
+        user: process.env.EMAIL_USER,
+        pass: process.env.EMAIL_PASS,
+      },
+    });
+    console.log("Transporter created.");
+
+    const resetLink = `${process.env.BASE_URL}/reset-password?token=${resetPassToken}`;
+    console.log("reset link created.");
+
+    const mailOptions = {
+      from: process.env.EMAIL_USER,
+      to: user.email,
+      subject: "><> Password Reset Request",
+      text: `Click the link to reset your password: ${resetLink}. This link will expire in 1 hour.`,
+    };
+    console.log("Mail options created.");
+
+    console.log("Sending email through transporter...");
+    await transporter.sendMail(mailOptions);
```

Comment:
this call is erroring out:
Error during password reset request: Error: Invalid login: 535-5.7.8 Username and Password not accepted. For more information, go to
535 5.7.8  https://support.google.com/mail/?p=BadCredentials 6a1803df08f44-6d8da9fe7f5sm8021006d6.80 - gsmtp